Port windows native-code to NAPI

zawata opened the pull request:
I'm shocked by how easy this was, took me less than a day to port, and worked the first try.

Interface might need some work but I've tested the following cases:

|OS arch|node arch|target process arch|result|
|-|-|-|-|
|x86_64|x86_64|x86_64|Pass|
|x86_64|x86_64|x86|Pass|
|x86_64|x86|x86_64|Fail (Expected)|
|x86_64|x86|x86|Pass|
|arm64|arm64|x86_64|Pass|
|arm64|arm64|x86|Pass|
|arm64|x86_64|arm64|Pass|
|arm64|x86|arm64|Fail (Expected)|
|arm64|x86_64|x86|Pass|
|arm64|x86|x86|Pass|
|arm64|x86|x86_64|Fail (Expected)|

Doesn't seem to be any risk for context-awareness since the module instance stores all the references it uses.

There are some concerns with thread-safety when using GetModuleHandle, but these shouldn't apply to use. Per the docs:
https://learn.microsoft.com/en-us/windows/win32/api/libloaderapi/nf-libloaderapi-getmodulehandlew#remarks

>The returned handle is not global or inheritable. It cannot be duplicated or used by another process.

We aren't sharing the handle to the module between contexts.

> The GetModuleHandle function returns a handle to a mapped module without incrementing its reference count. However, if this handle is passed to the [FreeLibrary](https://learn.microsoft.com/en-us/windows/desktop/api/libloaderapi/nf-libloaderapi-freelibrary) function, the reference count of the mapped module will be decremented. Therefore, do not pass a handle returned by GetModuleHandle to the FreeLibrary function. Doing so can cause a DLL module to be unmapped prematurely.
>
> This function must be used carefully in a multithreaded application. There is no guarantee that the module handle remains valid between the time this function returns the handle and the time it is used. For example, suppose that a thread retrieves a module handle, but before it uses the handle, a second thread frees the module. If the system loads another module, it could reuse the module handle that was recently freed. Therefore, the first thread would have a handle to a different module than the one intended.

At no point are we attempting to free the handle. It's a pointer to `ntdll` which should never go missing. We are likely to have much bigger issues if it does.

I'll start verifying this in electron but I wanted to get this up to solicit opinions.

zawata committed:
port to napi

zawata committed:
bump node-gyp

zawata committed:
require addon more saferer

ianhattendorf commented on src/readCwd.cpp:
```diff
@@ -0,0 +1,263 @@
+
+  void initializeNtCall() {
+    this->nt.p_ntdll = GetModuleHandleW(L"ntdll.dll");
+    this->nt.close = nt_close(GetProcAddress(this->nt.p_ntdll, "NtClose"));
```
This syntax is a little odd to me and hard to parse, I would expect something like this instead for these:

```suggestion
    this->nt.close = (nt_close) GetProcAddress(this->nt.p_ntdll, "NtClose");
```

ianhattendorf replied:
Also we should probably be checking the return value for `nullptr`.

zawata replied:
The code is more or less copied and pasted form the old version, but you're right, we should be doing some better error handling to prevent from seg-faulting the node/electron process.

julianmesa-gitkraken reviewed (approved):
Seems pretty direct conversion to NAPI.
Tested and works fine.
The code is the same from old version, so I do no expect any problem.

zawata committed:
better error handling

enable throwing c++ errors
actually check module handle and function pointers.
clean-up argument checks, support number-coercion
make failures more verbose
